openedx/core/djangoapps/course_date_signals/tests.py

```diff
@@ -114,5 +116,267 @@ def test_sequence_with_graded_and_ungraded_assignments(self):
+                ItemFactory.create(category='problem', parent=vertical2)
+
+                chapter3 = ItemFactory.create(category='chapter', parent=course5)
+                sequential3 = ItemFactory.create(category='sequential', parent=chapter3, graded=True)
+                vertical3 = ItemFactory.create(category='vertical', parent=sequential3)
+                ItemFactory.create(category='problem', parent=vertical3)
+
+                chapter4 = ItemFactory.create(category='chapter', parent=course5)
+                sequential4 = ItemFactory.create(category='sequential', parent=chapter4, graded=True)
+                vertical4 = ItemFactory.create(category='vertical', parent=sequential4)
+                ItemFactory.create(category='problem', parent=vertical4)
+            self.courses.append(course5)
+
+    @override_waffle_flag(CUSTOM_PLS, active=True)
+    def test_extract_dates_from_course_inheritance(self):
+        """
+        extract_dates_from_course should return a list of (block item location, field metadata dictionary)
+        and the blocks should inherit the dates from those above in the hiearchy
+        (ex. If a subsection is assigned a due date, its children should also have the same due date)
+        """
+        # course 1: With due_num_weeks but without any units. The subsection should have a set due date,
+        # while section should have the minimum number of weeks to complete as the due date (4 weeks).
+        course = self.courses[0]
+        chapter = course.get_children()[0]
+        sequential = chapter.get_children()[0]
+        expected_dates = [
+            (course.location, {}),
+            (chapter.location, timedelta(days=28)),
+            (sequential.location, {'due': timedelta(days=21)})
+        ]
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
+
+        # With due_num_weeks and a unit. The unit should inherit the due date.
+        vertical = ItemFactory.create(category='vertical', parent=sequential)
+        expected_dates.append((vertical.location, {'due': timedelta(days=21)}))
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
+
+        # With due_num_weeks and a problem. The problem should inherit the due date.
+        problem = ItemFactory.create(category='problem', parent=vertical)
+        expected_dates.append((problem.location, {'due': timedelta(days=21)}))
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
+
+    @override_waffle_flag(CUSTOM_PLS, active=True)
+    def test_extract_dates_from_course_custom_and_default_pls(self):
+        """
+        extract_dates_from_course should have the right set due dates for
+        subsections with custom PLS and subsections with the default PLS logic
```

Comment:
I think we can improve this docstring a little to explain that if the custom pls is not set, the subsection will fall back to the default PLS logic of evenly spaced sections